File: lexer/lexer.cpp
#include "../include/dotFun/lexer/lexer.h"

#include <cctype>

namespace dotFun {
    bool Lexer::match(char expected)
    {
        if (isAtEnd()) return false;
        if (m_source[m_current] != expected) return false;

        m_current++;
        m_column++;
        return true;
    }

    void Lexer::error(const std::string& message) {
        std::cerr << "[Line " << m_line << ", Column " << m_column << "] Lexer error: " << message << std::endl;
    }

    void Lexer::charLiteral()
    {
        if (peek() == '\\')
        {
            advance();
            advance();
        }
        else if (!isAtEnd())
        {
            advance();
        }

        if (match('\''))
        {
            std::string text = m_source.substr(m_start + 1, m_current - m_start - 2);
            addToken(dotFun::TokenType::CHAR_LITERAL, text);
        }
        else
        {
            error("Unterminated character literal. Expected closing single quote.");
        }
    }

    Lexer::Lexer(const std::string& source)
        : m_source(source) {
    }

    std::vector<dotFun::Token> Lexer::tokenize() {
        m_tokens.clear();
        m_start = 0;
        m_current = 0;
        m_line = 1;
        m_column = 1;

        while (!isAtEnd())
        {
            m_start = m_current;
            scanToken();
        }

        m_tokens.emplace_back(dotFun::TokenType::EOF_TOKEN, "", "", m_line, m_column);
        return m_tokens;
    }

    bool Lexer::isAtEnd() const {
        return m_current >= m_source.size();
    }

    char Lexer::advance() {
        char c = m_source[m_current++];
        if (c == '\n')
        {
            m_line++;
            m_column = 1;
        }
        else
        {
            m_column++;
        }
        return c;
    }

    void Lexer::addToken(dotFun::TokenType type) {
        addToken(type, "");
    }

    void Lexer::addToken(dotFun::TokenType type, const std::string& literal) {
        std::string text = m_source.substr(m_start, m_current - m_start);
        m_tokens.emplace_back(type, text, literal, m_line, m_column);
    }

    char Lexer::peek() const {
        if (isAtEnd()) return '\0';
        return m_source[m_current];
    }

    char Lexer::peekNext() const {
        if (m_current + 1 >= m_source.size()) return '\0';
        return m_source[m_current + 1];
    }

    void Lexer::scanToken() {
        skipWhitespace();

        if (isAtEnd()) return;

        char c = advance();

        switch (c)
        {
            // Single-character tokens
            case '(': addToken(dotFun::TokenType::LEFT_PAREN); break;
            case ')': addToken(dotFun::TokenType::RIGHT_PAREN); break;
            case '{': addToken(dotFun::TokenType::LEFT_BRACE); break;
            case '}': addToken(dotFun::TokenType::RIGHT_BRACE); break;
            case '[': addToken(dotFun::TokenType::LEFT_BRACKET); break;
            case ']': addToken(dotFun::TokenType::RIGHT_BRACKET); break;
            case ',': addToken(dotFun::TokenType::COMMA); break;
            case '.': addToken(dotFun::TokenType::DOT); break;
            case ';': addToken(dotFun::TokenType::SEMICOLON); break;
            case '*': addToken(dotFun::TokenType::STAR); break;
            case '%': addToken(dotFun::TokenType::PERCENT); break;

            case '!':
                addToken(match('=') ? dotFun::TokenType::NOT_BANG : dotFun::TokenType::NOT);
                break;

            case '=':
                addToken(match('=') ? dotFun::TokenType::EQUAL_EQUAL : dotFun::TokenType::EQUAL);
                break;

            case '&':
                if (match('&'))
                    addToken(dotFun::TokenType::AND_AND);
                else
                    addToken(dotFun::TokenType::AND); // single '&' as AND? Up to your spec
                break;

            case '|':
                if (match('|'))
                    addToken(dotFun::TokenType::OR_OR);
                else
                    addToken(dotFun::TokenType::OR); // single '|' as OR? Up to your spec
                break;

            case ':':
                addToken(dotFun::TokenType::COLON);
                break;

            case '"':
                stringLiteral();
                break;

            case '+':
                addToken(match('+') ? dotFun::TokenType::PLUS_PLUS : dotFun::TokenType::PLUS);
                break;

            case '-':
                addToken(match('-') ? dotFun::TokenType::MINUS_MINUS : dotFun::TokenType::MINUS);
                break;

            case '>':
                addToken(match('=') ? dotFun::TokenType::GREATER_EQUAL : dotFun::TokenType::GREATER);
                break;

            case '<':
                addToken(match('=') ? dotFun::TokenType::LESS_EQUAL : dotFun::TokenType::LESS);
                break;
            case '/':
                if (match('/'))
                {
                    while (peek() != '\n' && !isAtEnd())
                        advance();
                }
                else if (match('*'))
                {
                    while (!(peek() == '*' && peekNext() == '/') && !isAtEnd())
                    {
                        if (peek() == '\n') m_line++;
                        advance();
                    }

                    if (isAtEnd())
                    {
                        error("Unterminated block comment.");
                    }
                    else
                    {
                        advance();
                        advance();
                    }
                }
                else
                {
                    addToken(dotFun::TokenType::SLASH);
                }
                break;

            default:
                if (std::isdigit(c))
                {
                    numberLiteral();
                }
                else if (std::isalpha(c) || c == '_')
                {
                    identifier();
                }
                else
                {
                }
                break;
        }
    }

    void Lexer::stringLiteral()
    {
        while (peek() != '"' && !isAtEnd())
        {
            if (peek() == '\n') m_line++;
            advance();
        }

        if (isAtEnd())
        {
            return;
        }

        advance();

        std::string value = m_source.substr(m_start + 1, m_current - m_start - 2);
        addToken(dotFun::TokenType::STRING_LITERAL, value);
    }

    void Lexer::numberLiteral()
    {
        while (std::isdigit(peek()))
            advance();

        // Look for fractional part
        if (peek() == '.' && std::isdigit(peekNext()))
        {
            advance(); // consume '.'

            while (std::isdigit(peek()))
                advance();
        }

        std::string numStr = m_source.substr(m_start, m_current - m_start);
        addToken(dotFun::TokenType::NUMBER_LITERAL, numStr);
    }

    void Lexer::identifier()
    {
        while (std::isalnum(peek()) || peek() == '_')
            advance();

        std::string text = m_source.substr(m_start, m_current - m_start);

        auto typeOpt = keywordType(text);
        if (typeOpt.has_value())
        {
            addToken(typeOpt.value());
        }
        else
        {
            addToken(dotFun::TokenType::IDENTIFIER);
        }
    }

    void Lexer::skipWhitespace()
    {
        while (true)
        {
            char c = peek();
            switch (c)
            {
                case ' ':
                case '\r':
                case '\t':
                    advance();
                    break;

                case '\n':
                    advance();
                    break;
                default:
                    return;
            }
        }
    }

    std::optional<dotFun::TokenType> Lexer::keywordType(const std::string& text)
    {
        static const std::unordered_map<std::string, dotFun::TokenType> keywords = {
            {"class", dotFun::TokenType::CLASS},
            {"interface", dotFun::TokenType::INTERFACE},
            {"extends", dotFun::TokenType::EXTENDS},
            {"import", dotFun::TokenType::IMPORT},
            {"export", dotFun::TokenType::EXPORT},
            {"enum", dotFun::TokenType::ENUM},
            {"struct", dotFun::TokenType::STRUCT},
            {"public", dotFun::TokenType::PUBLIC},
            {"protected", dotFun::TokenType::PROTECTED},
            {"private", dotFun::TokenType::PRIVATE},
            {"if", dotFun::TokenType::IF},
            {"else", dotFun::TokenType::ELSE},
            {"elif", dotFun::TokenType::ELIF},
            {"while", dotFun::TokenType::WHILE},
            {"for", dotFun::TokenType::FOR},
            {"break", dotFun::TokenType::BREAK},
            {"continue", dotFun::TokenType::CONTINUE},
            {"async", dotFun::TokenType::ASYNC},
            {"fun", dotFun::TokenType::FUN},
            {"return", dotFun::TokenType::RETURN},
            {"true", dotFun::TokenType::TRUE},
            {"false", dotFun::TokenType::FALSE},
            {"nil", dotFun::TokenType::NIL},
            {"number", dotFun::TokenType::NUMBER},
            {"string", dotFun::TokenType::STRING},
            {"bool", dotFun::TokenType::BOOLEAN},
            {"char", dotFun::TokenType::CHAR},
            {"let", dotFun::TokenType::LET},
            {"val", dotFun::TokenType::VAL},
            {"global", dotFun::TokenType::GLOBAL},
            {"and", dotFun::TokenType::AND},
            {"or", dotFun::TokenType::OR},
            {"not", dotFun::TokenType::NOT},
            {"is", dotFun::TokenType::IS},
            {"in", dotFun::TokenType::IN},
            {"try", dotFun::TokenType::TRY},
            {"catch", dotFun::TokenType::CATCH},
            {"finally", dotFun::TokenType::FINALLY},
            {"throw", dotFun::TokenType::THROW},
            {"turn", dotFun::TokenType::TURN},
            {"case", dotFun::TokenType::CASE},
            {"default", dotFun::TokenType::DEFAULT},
            {"override", dotFun::TokenType::OVERRIDE},
            {"this", dotFun::TokenType::THIS},
            {"await", dotFun::TokenType::AWAIT},
            {"new", dotFun::TokenType::NEW}
        };

        auto it = keywords.find(text);
        if (it != keywords.end())
            return it->second;

        return std::nullopt;
    }
}

File: internal/lexer/lexer.go
package lexer

import (
	"fmt"
)

type Lexer struct {
	source  string
	start   int
	current int
	line    int
	column  int
	tokens  []Token
}

func NewLexer(source string) *Lexer {
	return &Lexer{
		source: source,
		line:   1,
		column: 1,
		tokens: []Token{},
	}
}

func (l *Lexer) Lex() ([]Token, error) {
	for !l.isAtEnd() {
		l.start = l.current
		err := l.scanToken()
		if err != nil {
			return nil, err
		}
	}
	l.tokens = append(l.tokens, Token{
		Type:   EOF_TOKEN,
		Lexeme: "",
		Line:   l.line,
		Column: l.column,
	})
	return l.tokens, nil
}

func (l *Lexer) scanToken() error {
	c := l.advance()

	switch c {
	case ' ', '\r', '\t':
		return nil
	case '\n':
		l.line++
		l.column = 0
		return nil

	case '/':
		if l.match('/') {
			for !l.isAtEnd() && l.peek() != '\n' {
				l.advance()
			}
			return nil
		} else if l.match('*') {
			for {
				if l.isAtEnd() {
					return fmt.Errorf("Unterminated block comment at line %d", l.line)
				}
				if l.peek() == '*' && l.peekNext() == '/' {
					l.advance()
					l.advance()
					break
				}
				if l.peek() == '\n' {
					l.line++
					l.column = 0
				}
				l.advance()
			}
			return nil
		}
		return l.addToken(SLASH)

	case '"':
		return l.string()

	case '\'':
		return l.string()

	default:
		if isDigit(c) {
			return l.number()
		}
		if isAlpha(c) || c == '_' {
			return l.identifier()
		}
	}

	switch c {
	case '+':
		if l.match('+') {
			return l.addToken(PLUS_PLUS)
		}
		return l.addToken(PLUS)
	case '-':
		if l.match('-') {
			return l.addToken(MINUS_MINUS)
		}
		if l.match('>') {
			return l.addToken(ARROW)
		}
		return l.addToken(MINUS)
	case '*':
		return l.addToken(STAR)
	case '%':
		return l.addToken(PERCENT)
	case '=':
		if l.match('=') {
			return l.addToken(EQUAL_EQUAL)
		}
		return l.addToken(EQUAL)
	case '!':
		if l.match('=') {
			return l.addToken(NOT_BANG)
		}
		return l.addToken(NOT_BANG)
	case '&':
		if l.match('&') {
			return l.addToken(AND_AND)
		}
		return l.addToken(BIT_AND)
	case '|':
		if l.match('|') {
			return l.addToken(OR_OR)
		}
		return l.addToken(BIT_OR)
	case '>':
		if l.match('=') {
			return l.addToken(GREATER_EQUAL)
		}
		if l.match('>') {
			return l.addToken(SHIFT_RIGHT)
		}
		return l.addToken(GREATER)
	case '<':
		if l.match('=') {
			return l.addToken(LESS_EQUAL)
		}
		if l.match('<') {
			return l.addToken(SHIFT_LEFT)
		}
		return l.addToken(LESS)
	case ':':
		if l.match(':') {
			return l.addToken(COLON_COLON)
		}
		return l.addToken(COLON)
	case '?':
		return l.addToken(QUESTION)
	case '.':
		if l.match('.') && l.match('.') {
			return l.addToken(ELLIPSIS)
		}
		return l.addToken(DOT)
	case ',':
		return l.addToken(COMMA)
	case ';':
		return l.addToken(SEMICOLON)
	case '(':
		return l.addToken(LEFT_PAREN)
	case ')':
		return l.addToken(RIGHT_PAREN)
	case '{':
		return l.addToken(LEFT_BRACE)
	case '}':
		return l.addToken(RIGHT_BRACE)
	case '[':
		return l.addToken(LEFT_BRACKET)
	case ']':
		return l.addToken(RIGHT_BRACKET)
	case '$':
		return l.addToken(DOLLAR)
	default:
		return fmt.Errorf("Unexpected character '%c' at line %d column %d", c, l.line, l.column)
	}

	return nil
}

func (l *Lexer) addToken(t TokenType) error {
	text := l.source[l.start:l.current]
	l.tokens = append(l.tokens, Token{
		Type:   t,
		Lexeme: text,
		Line:   l.line,
		Column: l.column,
	})
	return nil
}

func (l *Lexer) advance() byte {
	c := l.source[l.current]
	l.current++
	l.column++
	return c
}

func (l *Lexer) match(expected byte) bool {
	if l.isAtEnd() {
		return false
	}
	if l.source[l.current] != expected {
		return false
	}
	l.current++
	l.column++
	return true
}

func (l *Lexer) peek() byte {
	if l.isAtEnd() {
		return 0
	}
	return l.source[l.current]
}

func (l *Lexer) peekNext() byte {
	if l.current+1 >= len(l.source) {
		return 0
	}
	return l.source[l.current+1]
}

func (l *Lexer) isAtEnd() bool {
	return l.current >= len(l.source)
}

func (l *Lexer) string() error {
	for {
		if l.isAtEnd() {
			return fmt.Errorf("Unterminated string literal at line %d", l.line)
		}
		if l.peek() == '"' {
			l.advance()
			break
		}
		if l.peek() == '\n' {
			l.line++
			l.column = 0
		}
		l.advance()
	}
	return l.addToken(STRING_LITERAL)
}

func (l *Lexer) number() error {
	for isDigit(l.peek()) {
		l.advance()
	}

	if l.peek() == '.' && isDigit(l.peekNext()) {
		l.advance()

		for isDigit(l.peek()) {
			l.advance()
		}
	}

	return l.addToken(NUMBER_LITERAL)
}

func (l *Lexer) identifier() error {
	for isAlphaNumeric(l.peek()) || l.peek() == '_' {
		l.advance()
	}

	text := l.source[l.start:l.current]

	tokType := lookupKeyword(text)
	return l.addToken(tokType)
}

func lookupKeyword(ident string) TokenType {
	switch ident {
	case "class":
		return CLASS
	case "interface":
		return INTERFACE
	case "extends":
		return EXTENDS
	case "import":
		return IMPORT
	case "export":
		return EXPORT
	case "enum":
		return ENUM
	case "struct":
		return STRUCT
	case "public":
		return PUBLIC
	case "protected":
		return PROTECTED
	case "private":
		return PRIVATE
	case "override":
		return OVERRIDE
	case "this":
		return THIS
	case "new":
		return NEW
	case "super":
		return SUPER
	case "constructor":
		return CONSTRUCTOR
	case "data":
		return DATA
	case "instanceof":
		return INSTANCEOF
	case "if":
		return IF
	case "else":
		return ELSE
	case "elif":
		return ELIF
	case "while":
		return WHILE
	case "for":
		return FOR
	case "break":
		return BREAK
	case "continue":
		return CONTINUE
	case "async":
		return ASYNC
	case "await":
		return AWAIT
	case "fun":
		return FUN
	case "return":
		return RETURN
	case "true":
		return TRUE
	case "false":
		return FALSE
	case "nil":
		return NIL
	case "let":
		return LET
	case "val":
		return VAL
	case "global":
		return GLOBAL
	case "and":
		return AND
	case "or":
		return OR
	case "not":
		return NOT
	case "is":
		return IS
	case "in":
		return IN
	case "of":
		return OF
	case "try":
		return TRY
	case "catch":
		return CATCH
	case "finally":
		return FINALLY
	case "throw":
		return THROW
	case "turn":
		return TURN
	case "case":
		return CASE
	case "default":
		return DEFAULT
	default:
		return IDENTIFIER
	}
}

func isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

func isAlpha(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
}

func isAlphaNumeric(c byte) bool {
	return isAlpha(c) || isDigit(c)
}
